core/app/models/spree/promotion.rb

```diff
@@ -126,9 +126,10 @@ def activate(order:, line_item: nil, user: nil, path: nil, promotion_code: nil)
     # called anytime order.recalculate happens
     def eligible?(promotable, promotion_code: nil)
       return false if inactive?
-      return false if usage_limit_exceeded?
-      return false if promotion_code && promotion_code.usage_limit_exceeded?
+      return false if usage_limit_exceeded?(excluded_orders: [promotable])
```

Comment:
I think we must handle here the different cases a `promotable` object can be:
- `Spree::Order`
- `Spree::LineItem`
- `Spree::Shipment`

You can simulate these cases by creating different promos with different [_Action_' adjustment types.](https://cedum.s3-eu-west-2.amazonaws.com/yfzqwsbz4).

Reply:
I had this doubt but I thought that I could ignore LineItems and Shipments when computing eligibility check on them because usage count was not directly affected.

Do you think that updating`#eligible?`to handle non Order promotables excluding Orders to which LineItems and Shipments belong to would be enough?

Reply:
> Do you think that updating `#eligible?` to handle non Order promotables excluding Orders to which LineItems and Shipments belong to would be enough?

👍 yes. AFAIK, a promotion's usage count is always the number of distinct orders that promo is applied. I cannot think of cases where it's counted by a different object.